### PR TITLE
[DRAFT] Error correction for BIP-39 mnemonic phrases

### PR DESCRIPTION
**Pull request by nategraf**
Introduce a draft for proposal for error correction of BIP-39 mnemonic phrases.

Direct discussion to https://github.com/celo-org/celo-proposals/issues/225

### CODE REVIEW

File: CIPs/cip-draft_mnemonic_error_correcting_codes.md [hunk 1]
```diff
@@ -0,0 +1,233 @@
...
+* A "quorum" subset of symbols (i.e. points) are consistent with some `K-1` degree polynomial.
+* And the re-encoded codeword contains a valid BIP-39 checksum.
+
+<!-- TOOD(victor) Refactor the math in this section. It is _not_ easy to read -->
+
+In this case, the quorum of points required is the number of points available minus the number of
+acceptable unknown errors :`(N - e) - floor((N - K - e) / 2) = ceil((N + K - e) / 2)`. Given a
+particular quorum, it is consistent with some `K-1` degree polynomial if and only if, taking any `K`
+points, the remaining `ceil((N - K - e) / 2)` points are consistent with the polynomial interpolated
+from those `K` points. When  the points are uniformly random and independent of each other the
```
Comment (kobigurk) [inline CIPs/cip-draft_mnemonic_error_correcting_codes.md:189]: Maybe easier to phrase - once you interpolate the polynomial, the other evaluations are uniquely determined, leading to a 2^-11 probability of being correct, and the probability you mentioned overall.

File: CIPs/cip-draft_mnemonic_error_correcting_codes.md [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+probability that this is true is `y = 2 ^ (-11 * ceil((N - K - e) / 2))`. (11 because the points are
+in `GF(2^11)`)
+
+In a given phrase, there are `(N - e) choose ceil((N + K - e) / 2)` quorums. When the words of the
+phrase are chosen independently and at random, The overall chance of decoding a plain BIP-39 phrase
+is `1 - (1 - y) ^ ((N - e) choose ceil((N + K - e) / 2))`, with `y` defined above.
```
Comment (kobigurk) [inline CIPs/cip-draft_mnemonic_error_correcting_codes.md:195]: Nice to add - "meaning, at least one succeeds".

File: CIPs/cip-draft_mnemonic_error_correcting_codes.md [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+a multiple of 3 between 12 and 24 inclusive (i.e. It must be a valid number of words for BIP-39).
+`K` MUST be a multiple of 3 between 12 and 24 inclusive, less than or equal to `N`. Note that when
+`N = K`, no error correction is provided and the scheme is equivalent to BIP-39 phrase with `N`
+words.
+
+With a given selection of `N` and `K`, it will be possible to correct `N-K` symbols at known
+locations (i.e. when the provided token is not a valid BIP-39 word, or is indicated as missing) and
+up to `floor(N-K/2)` symbols at known locations (e.g. when word orderings are swapped, or replaced
+with another valid BIP-39 word). This is a result of the
+[Singleton bound](https://en.wikipedia.org/wiki/Singleton_bound).
```
[the remark this answers is not in the document]
Comment (nategraf) [reply CIPs/cip-draft_mnemonic_error_correcting_codes.md]: Removed that sentance

File: CIPs/cip-draft_mnemonic_error_correcting_codes.md [hunk 4]
```diff
@@ -0,0 +1,233 @@
...
+
+| N  | K  | entropy (bits) | correctable errors (words) | correctable erasures (words) |
```
[the remark this answers is not in the document]
Comment (nategraf) [reply CIPs/cip-draft_mnemonic_error_correcting_codes.md:101]: Added a note about the word erasure.

File: CIPs/cip-draft_mnemonic_error_correcting_codes.md [hunk 5]
```diff
@@ -0,0 +1,233 @@
...
+| 18 | 15 |            159 |                          1 |                            3 |
+| 21 | 12 |            125 |                          4 |                            9 |
+| 21 | 15 |            158 |                          3 |                            6 |
+| 21 | 18 |            191 |                          1 |                            3 |
+| 24 | 12 |            124 |                          6 |                           12 |
+| 24 | 15 |            157 |                          4 |                            9 |
+| 24 | 18 |            190 |                          3 |                            6 |
+| 24 | 21 |            223 |                          1 |                            3 |
+
+Note that the entropy available for a given `N` and `K` is slightly less (up to 4 bits) than a
```
[the remark this answers is not in the document]
Comment (nategraf) [reply CIPs/cip-draft_mnemonic_error_correcting_codes.md:114]: Added an extended description of why there is a reduction in entropy.

File: CIPs/cip-draft_mnemonic_error_correcting_codes.md (hunk 2 above)
Comment (nategraf) [reply CIPs/cip-draft_mnemonic_error_correcting_codes.md]: Added a clause to that effect

File: CIPs/cip-draft_mnemonic_error_correcting_codes.md (hunk 1 above)
Comment (nategraf) [reply CIPs/cip-draft_mnemonic_error_correcting_codes.md]: Rephrased this paragraph. Let me know what you think.

**Comment by nategraf** [comment]
In 02188218b83478407c861e979ba047991166237e, I went ahead and removed the CIP on model-based error correction. After realizing some mistakes I'd made in the formulation, and not being able to fix them in a particularly satisfying way given the amount of time I want to spend on this, I check-pointed by work in the branch [victor/model-based-phrase-correction](https://github.com/celo-org/celo-proposals/compare/victor/model-based-phrase-correction) and decided to remove that CIP from this pull request. Anyone interested in that approach should feel free to revive it.

**Review by YazzyYaz — APPROVED** [review]

File: CIPs/cip-39.md [hunk 6]
```diff
@@ -0,0 +1,250 @@
...
+happens, it can be a catastrophic event for the user, losing any assets or identity held by accounts
+derived from the phrase.
+
+In the case that the user has a copy of their phrase with errors, introduced when copying the phrase
+to paper, during storage (e.g. water damage, dog nibbled on it), or when typing the phrase into a
+wallet application, the raw data (i.e. entropy) may still be available in the key, but has been
+obscured by the introduced errors. Given expertise and time, a user may being able to manually
+locate and remove any errors present in the phrase. Most users don't have the required expertise or
+time. Hueristic-based techinques, such as suggesting phrases that have a low edit distance from the
+origonal, provide an option to correct phrases with common errors such as typos or substitution for
```
Comment (carterqw2) [inline CIPs/cip-39.md:50]:
```suggestion
original, provide an option to correct phrases with common errors such as typos or substitution for
```

File: CIPs/cip-39.md [hunk 7]
```diff
@@ -0,0 +1,250 @@
...
+simmilar words. https://github.com/celo-org/celo-monorepo/pull/8034 provides an example of how to do
+this.
+
+When information is lost from the phrase (e.g. some of the words are entirely missing), or when the
+heuristic does not do appropriately model a particular error, heuristics based approaches may fail.
+Proposed here is the introduction of error correcting codes, which can recover phases with a number
```
Comment (carterqw2) [inline CIPs/cip-39.md:56]:
```suggestion
Proposed here is the introduction of error correcting codes, which can recover phrases with a number
```

File: CIPs/cip-39.md [hunk 8]
```diff
@@ -0,0 +1,250 @@
...
+of arbitrary errors, covering many cases that model based approaches cannot.
+
+Compared to heursistc based methods, this method can handle a larger class of errors without making
```
Comment (carterqw2) [inline CIPs/cip-39.md:59]:
```suggestion
Compared to heuristic based methods, this method can handle a larger class of errors without making
```

**Pull request merged**